index.bs

```diff
@@ -73,34 +73,129 @@ command/response format of WebDriver, this permits events to stream
+on the [=remote end=], and the <dfn>local end schema</dfn> which defines the
+format of messages produced on the [=remote end=] and consumed on the [=local
+end=]
+
+## Schema ## {#protocol-schema}
+
+Issue: Should this be an appendix?
+
+This section gives the initial contents of the [=remote end schema=] and
+[=local end schema=]. These are augmented by the schema fragments defined in
+the remainder of the specification.
+
+[=Remote end schema=]
+
+```
+Command = {
+  id: uint,
+  CommandData,
+}
+
+CommandData = (
+  SessionCommand
+)
+
+EmptyParams = { *text }
+```
+
+[=Local end schema=]
+
+```
+Message = (
+  CommandResponse //
+  Event
+)
 
-Note: These definitions do not form any normative requirements, rather
-they are used elsewhere as part of such requirements.
+CommandResponse = (
+  id: uint,
+  ResponseData,
+)
+
+ResponseData = (
+  Error //
+  CommandResult
+)
+
+Error = (
+  id: uint,
+  error: (
+    "unknown error" /
+    "unknown method" /
+    "invalid argument"
+  ),
+  message: text,
+  stacktrace: text,
+)
+
+CommandResult = {
+  value: ResultData,
+}
+
+ResultData = (
+  EmptyResult //
+  SessionResult
+)
+
+EmptyResult = { *text }
+
+Event = ()
+```
+
+## Session ## {#session}
+
+WebDriver BiDi uses the same [=/session=] concept as WebDriver.
 
 ## Modules ## {#protocol-modules}
 
-The WebDriver BiDi protocol is organized into modules. A
-<dfn>module</dfn> is defined by:
- - A unique string <dfn export for=module>name</dfn>, implicit in its
-   definition.
- - A set of [=commands=].
- - A set of [=events=].
+The WebDriver BiDi protocol is organized into modules.
 
-Each [=module=] represents a collection of related [=commands=] and
-[=events=] pertaining to a certain aspect of the user agent. For
-example, a module might contain functionality for inspecting and
+Each <dfn export for=module>module</dfn> represents a collection of related
+[=commands=] and [=events=] pertaining to a certain aspect of the user
+agent. For example, a module might contain functionality for inspecting and
 manipulating the DOM, or for script execution.
 
+Each module has a <dfn>module name</dfn> which is a string literal. The
+[=command name=] and [=event name=] for commands and events defined in the
+module start with the [=module name=] followed by a period <code>"."</code>.
+
+Modules which contain [=commands=] define [=remote end schema=]
+fragments. These provide choices in the <code>CommandData</code> group for the
+module's [=commands=],and can also define additional schema properties. They
+can also define [=local end schema=] fragments that provide additional choices
+in the <code>ResultData</code> group for the results of commands in the module.
+
+Modules which contain events define [=local end schema=] fragments that are
+choices in the <code>Event</code> group for the module's [=events=].
+
+An implementation may define <dfn>extension modules</dfn>. These must have a
```

Comment:
How about extension commands? Does the setup we have allow for that? (I ask because of extension specs, not vendor-specific stuff.)

Reply:
The point is supposed to be that extension modules can contain extension commands.

Reply:
In other words, we're not allowing for extension commands or extension events in modules that are defined by this specification?

I don't particularly mind at this point because we're a long way from adding anything vendor-specific. It seems to me we'd regret it like all other prefixing. Can we maybe have an open issue and a link to in the spec, to steer the first implementer tempted to do something towards that issue, where we can then work it out with a real case to solve?

Reply:
Modules are mostly a spec-organisation convenience; I think the only way they really leak into the protocol is in the way we enable events. So I don't think there's a *lot* of user-facing difference between allowing a `chrome:runtime` module with a `doSomethingChromeish` command and having the `runtime` module be extended with a `chrome:doSomethingChromeish` command. But I think organisationally it's clearer that everything in that module is vendor specific and it means that you always need to explicitly opt-in to the vendor specfic things; you can just subscribe to all `runtime` events and start depending on the browser-specific events you get back.

For WebDriver/HTTP we already have vendor extensions and it's not totally unproblematic, but it's nowhere near as problematic as e.g. prefixes in CSS because mostly for authors it's abstracted away by libraries and there's only a few of those, and typically people writing automation control both ends of the pipeline i.e. they're writing the tests and picking the browsers they run in. Whilst I *certainly* think we want the base spec to be good enough for all common use cases to be cross-browser, I also think we're going to end up with some non-trivial amount of vendor-specific stuff (much of which might be a candidate for later standardisation).